mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala

```diff
@@ -41,9 +41,18 @@ import org.apache.spark.sql.DataFrame
  *                partition boundaries.
  */
 @Since("1.0.0")
-class BinaryClassificationMetrics @Since("1.3.0") (
-    @Since("1.3.0") val scoreAndLabels: RDD[(Double, Double)],
-    @Since("1.3.0") val numBins: Int) extends Logging {
+class BinaryClassificationMetrics @Since("3.0.0") (
+    @Since("1.3.0") val scoreAndLabelsWithOptWeight: RDD[_ <: Product],
```

Comment:
Let's see, so last time we decided this was OK because the type change is only visible at compile-time, and this should be source compatible. I don't _think_ the argument name change will be an issue here as it isn't a named optional arg. should be OK but we'll have to check with MiMa.

Reply:
BTW do we check elsewhere for positive weight? they need to be > 0?

Reply:
no we currently do not check for the weight > 0, should that live in BinaryClassificationMetrics?  I should probably update the other metrics classes as well then?

Reply:
There's no useful case for negative weights right? I can see allowing 0 for convenience, as it at least has a clear meaning.

Maybe this could be a separate change, but yeah anywhere the user supplies weights, it's useful to check this _somewhere_ along the way.

Reply:
Yes, I don't believe there is a useful case for negative weights but 0 may be useful.
I added a check in binary classification metrics.  Should I add to the other evaluators in this PR as well?

Reply:
specifically, the check was:
```
require(weight >= 0.0, s"instance weight, $weight has to be >= 0.0")
```